Cauldron/Assets/Scripts/ColoredByBrewedPotion.cs

```diff
@@ -0,0 +1,195 @@
+        [SerializeField] private Color placeboColor = Color.HSVToRGB(0.97f, 0.68f, 0.27f);
+        
+        [Range(-1f, 1f)] [SerializeField] private float saturation = 0f;
+        [Range(-1f, 1f)] [SerializeField] private float brightness = 1f;
+        
+        [Tooltip("Delay in seconds")]
+        [SerializeField] private float delay = 0f; 
+        
+        [SerializeField] private bool tween = false;  
+        
+        [Tooltip("Tween time in seconds")]
+        [SerializeField] private float tweenTime = 1f; 
+
+        private RecipeProvider recipeProvider;
+        private Cauldron cauldron;
+
+        [SerializeField] private SpriteRenderer spriteRenderer;
+        [SerializeField] private ParticleSystem particleSystem;
+
+
+        [Inject]
+        private void Construct(Cauldron cauldron, RecipeProvider recipeProvider)
+        {
+            this.recipeProvider = recipeProvider;
+            this.cauldron = cauldron;
+        }
+
+        private void OnValidate()
+        {
+            spriteRenderer = GetComponent<SpriteRenderer>();
+            particleSystem = GetComponent<ParticleSystem>();
+            if (!tween)
+            {
+                tweenTime = 0;
+            }
+        }
+
+        private void Start()
+        {
+            cauldron.PotionBrewed += DefineColor;
+            cauldron.PotionAccepted += ResetColor;
+            cauldron.PotionDeclined += PotionDeclined;
+        }
+
+        private void OnDestroy()
+        {
+            cauldron.PotionBrewed -= DefineColor;
+            cauldron.PotionAccepted -= ResetColor;
+            cauldron.PotionDeclined -= PotionDeclined;
+            StopAllCoroutines();
+        }
+
+        private void PotionDeclined()
+        {
+            ResetColor(Potions.Placebo);
+        }
+        
+        private void ResetColor(Potions potion)
+        {
+            if (spriteRenderer) 
+            { 
+                StartCoroutine(tween ? DoTweenColor(neutralColor, tweenTime) : SetColor(neutralColor)); 
+            } 
+            if (particleSystem) 
+            { 
+                StartCoroutine(tween ? DoTweenColor(neutralColor, tweenTime) : SetColor(neutralColor)); 
+            } 
+        }
+        
+        private void DefineColor(Potions potion)
+        {
+            Recipe recipe = recipeProvider.GetRecipeForPotion(potion);
+            Color color;
+            if (recipe != null)
+            {
+                color = recipe.color;
+            }
+            else
+            {
+                color = placeboColor;
+            }
+
+            if (spriteRenderer)
+            {
+                ChangeColor(color, spriteRenderer);
+            }
+
+            if (particleSystem)
+            {
+                ChangeColor(color, particleSystem);
+            }
+        }
+        
+        private void ChangeColor(Color color, Component component)
+        {
+            color = ModifyColor(color);
```

Comment:
You should stop current async process before starting a new one.
In case it's still running)